src/xdp/xsk.c

```diff
@@ -2947,52 +3082,50 @@ XskSockoptSetUmem(
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto Exit;
-    }
+    if (RequestorMode != KernelMode) {
+        Umem->Mapping.Mdl =
+            IoAllocateMdl(
+                Umem->Reg.Address,
+                (ULONG)Umem->Reg.TotalSize,
+                FALSE, // SecondaryBuffer
+                FALSE, // ChargeQuota
+                NULL); // Irp
+        if (Umem->Mapping.Mdl == NULL) {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Exit;
+        }
 
-    __try {
-        MmProbeAndLockPages(Umem->Mapping.Mdl, RequestorMode, IoWriteAccess);
-    } __except (EXCEPTION_EXECUTE_HANDLER) {
-        Status = GetExceptionCode();
-        goto Exit;
-    }
+        __try {
+            MmProbeAndLockPages(Umem->Mapping.Mdl, RequestorMode, IoWriteAccess);
+        } __except (EXCEPTION_EXECUTE_HANDLER) {
+            Status = GetExceptionCode();
+            goto Exit;
+        }
 
-    //
-    // MmGetSystemAddressForMdlSafe and MmMapLockedPagesSpecifyCache do not
-    // preserve large pages in system address mappings. Use the reserved MDL
-    // mapping routines to ensure large pages are propagated into the kernel.
-    // Note that the reserved mapping allocates the mapping size based on best-
-    // case page-aligned buffers, so account for the MDL offset, too.
-    //
-    Umem->ReservedMapping =
-        MmAllocateMappingAddress(
-            BYTE_OFFSET(Umem->Reg.Address) + Umem->Reg.TotalSize, POOLTAG_UMEM);
-    if (Umem->ReservedMapping == NULL) {
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto Exit;
-    }
+        //
+        // MmGetSystemAddressForMdlSafe and MmMapLockedPagesSpecifyCache do not
+        // preserve large pages in system address mappings. Use the reserved MDL
+        // mapping routines to ensure large pages are propagated into the kernel.
+        // Note that the reserved mapping allocates the mapping size based on best-
+        // case page-aligned buffers, so account for the MDL offset, too.
+        //
+        Umem->ReservedMapping =
+            MmAllocateMappingAddress(
+                BYTE_OFFSET(Umem->Reg.Address) + Umem->Reg.TotalSize, POOLTAG_UMEM);
+        if (Umem->ReservedMapping == NULL) {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Exit;
+        }
 
-    Umem->Mapping.SystemAddress =
-        MmMapLockedPagesWithReservedMapping(
-            Umem->ReservedMapping, POOLTAG_UMEM, Umem->Mapping.Mdl, MmCached);
-    if (Umem->Mapping.SystemAddress == NULL) {
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto Exit;
+        Umem->Mapping.SystemAddress =
+            MmMapLockedPagesWithReservedMapping(
+                Umem->ReservedMapping, POOLTAG_UMEM, Umem->Mapping.Mdl, MmCached);
+        if (Umem->Mapping.SystemAddress == NULL) {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto Exit;
+        }
+    } else {
+        Umem->Mapping.SystemAddress = Umem->Reg.Address;
```

Comment:
We may also want to support the kernel caller providing an MDL rather than a virtual address - that would be much more efficient if the caller happens to have an MDL on hand already.

Reply:
Makes sense. It seems like how we want to expose that functionality will impact whether or not it makes sense to add this to the API surface now. MsQuic does not have the MDL on hand so there is no immediate need for this.

Would exposing this as a v2 umem register IOCTL make the most sense? In which case, that can be added when needed and would only result in a minor version bump.

Reply:
@mtfriesen ^

Reply:
Yeah, deferring the feature until a need arises seems best.